src/k8s_upgrade.py

```diff
@@ -0,0 +1,497 @@
+            logger.debug(f"Resume refresh succeeded: {message}")
+
+
+class _PostUpgradeCheckMongoDB(EventBase):
+    """Run post upgrade check on MongoDB to verify that the cluster is healhty."""
+
+    def __init__(self, handle):
+        super().__init__(handle)
+
+
+class MongoDBUpgrade(GenericMongoDBUpgrade):
+    """Handlers for upgrade events."""
+
+    post_app_upgrade_event = EventSource(_PostUpgradeCheckMongoDB)
+    post_cluster_upgrade_event = EventSource(_PostUpgradeCheckMongoDB)
+
+    def __init__(self, charm: "MongoDBCharm"):
+        self.charm = charm
+        super().__init__(charm, PEER_RELATION_ENDPOINT_NAME)
+
+    @override
+    def _observe_events(self, charm: "MongoDBCharm") -> None:
+        self.framework.observe(
+            charm.on[PRECHECK_ACTION_NAME].action, self._on_pre_upgrade_check_action
+        )
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_created,
+            self._on_upgrade_peer_relation_created,
+        )
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_changed, self._reconcile_upgrade
+        )
+        self.framework.observe(charm.on[RESUME_ACTION_NAME].action, self._on_resume_upgrade_action)
+        self.framework.observe(self.post_app_upgrade_event, self.run_post_app_upgrade_task)
+        self.framework.observe(self.post_cluster_upgrade_event, self.run_post_cluster_upgrade_task)
+
+    def _reconcile_upgrade(self, _, during_upgrade: bool = False) -> None:
+        """Handle upgrade events."""
+        if not self._upgrade:
+            logger.debug("Peer relation not available")
+            return
+        if not self._upgrade.versions_set:
+            logger.debug("Peer relation not ready")
+            return
+        if self.charm.unit.is_leader() and not self._upgrade.in_progress:
+            # Run before checking `self._upgrade.is_compatible` in case incompatible upgrade was
+            # forced & completed on all units.
+            self.charm.version_checker.set_version_across_all_relations()
+            self._upgrade.set_versions_in_app_databag()
+
+        if self._upgrade.unit_state is UnitState.RESTARTING:  # Kubernetes only
+            if not self._upgrade.is_compatible:
+                logger.info(
+                    f"Refresh incompatible. If you accept potential *data loss* and *downtime*, you can continue with `{RESUME_ACTION_NAME} force=true`"
+                )
+                self.charm.status.set_and_share_status(Config.Status.INCOMPATIBLE_UPGRADE)
+                return
+        if not during_upgrade and self.charm.db_initialised and self.charm.is_db_service_ready():
+            self._upgrade.unit_state = UnitState.HEALTHY
+            self.charm.status.set_and_share_status(ActiveStatus())
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+
+        self._set_upgrade_status()
+
+    def _set_upgrade_status(self):
```

Comment:
maybe you can use this function where you set status in the `upgrade` function in `src/charm.py`

Reply:
As I said above: During a k8s upgrade since there's a restart, it puts the status to `MaintenanceStatus` during the restart.
I had a look, trying if I could handle it here, but it doesn't work well because `MaintenanceStatus` is used for different things and it would break some other parts of the charm.
I'll try to think about something better in order to handle that, but I think it's safe to set to `ActiveStatus`  in the upgrade-charm event since it will anyway be handled right after by the `_reconcile_upgrade`  and `post_app_upgrade_check`.